File: Templates/DisjointSetUnionByRank.py
class DisjointSet:
    def __init__(self,n):
        self.n = n
        self.parent = [i for i in range(n+1)]
        self.rank = [1]*(n+1)
    
    def getParent(self, node) -> int:
        if self.parent[node] == node:
            return node
        
        self.parent[node] = self.getParent(self.parent[node])
        return self.parent[node]


    def unionByRank(self, u, v):
        u = self.getParent(u)
        v = self.getParent(v)

        if u==v:
            return
        
        if self.rank[u] < self.rank[v]:
            self.parent[u] = v
        elif self.rank[v] < self.rank[u]:
            self.parent[v] = u
        else:
            self.parent[v] = u
            self.rank[u]+=1

File: Templates/DisjointSetUnionByRank.cpp
#include <bits/stdc++.h>
using namespace std;

class DisjointSet {
private:
    vector<int> parent, rank;
    int n;
public:
    DisjointSet(int n){
        this->n = n;
        for (int i=0;i<=n; ++i){
            parent.push_back(i);
            rank.push_back(1);
        }
    }

    int findParent(int x){
        if (parent[x] != x){
            parent[x] = findParent(parent[x]);
        }
        return parent[x];
    }

    void unionByRank(int x, int y){
        x = findParent(x);
        y = findParent(y);
        if (x == y) return;

        if (rank[x] < rank[y]){
            parent[x] = y;
        } else if (rank[x] > rank[y]){
            parent[y] = x;
        } else {
            parent[y] = x;
            rank[x]++;
        }
    }
};